File: smash-ranks-android/app/src/main/java/com/garpr/android/adapters/HomePagerAdapter.java
package com.garpr.android.adapters;

import android.support.annotation.Nullable;
import android.support.v4.util.SparseArrayCompat;
import android.support.v4.view.PagerAdapter;
import android.view.View;
import android.view.ViewGroup;

import com.garpr.android.activities.HomeActivity;
import com.garpr.android.misc.Searchable;
import com.garpr.android.views.FavoritePlayersLayout;
import com.garpr.android.views.RankingsLayout;
import com.garpr.android.views.SearchableFrameLayout;
import com.garpr.android.views.TournamentsLayout;

import java.lang.ref.WeakReference;

public class HomePagerAdapter extends PagerAdapter implements Searchable {

    private static final int POSITION_FAVORITE_PLAYERS = HomeActivity.POSITION_FAVORITE_PLAYERS;
    private static final int POSITION_RANKINGS = HomeActivity.POSITION_RANKINGS;
    private static final int POSITION_TOURNAMENTS = HomeActivity.POSITION_TOURNAMENTS;

    private final SparseArrayCompat<WeakReference<SearchableFrameLayout>> mPages;


    public HomePagerAdapter() {
        mPages = new SparseArrayCompat<>(getCount());
    }

    @Override
    public void destroyItem(final ViewGroup container, final int position, final Object object) {
        container.removeView((View) object);
        mPages.removeAt(position);
    }

    @Override
    public int getCount() {
        return 3;
    }

    @Override
    public Object instantiateItem(final ViewGroup container, final int position) {
        final SearchableFrameLayout view;

        switch (position) {
            case POSITION_FAVORITE_PLAYERS:
                view = FavoritePlayersLayout.inflate(container);
                break;

            case POSITION_TOURNAMENTS:
                view = TournamentsLayout.inflate(container);
                break;

            case POSITION_RANKINGS:
                view = RankingsLayout.inflate(container);
                break;

            default:
                throw new RuntimeException("illegal position: " + position);
        }

        container.addView(view);
        mPages.put(position, new WeakReference<>(view));

        return view;
    }

    @Override
    public boolean isViewFromObject(final View view, final Object object) {
        return view == object;
    }

    @Override
    public void search(@Nullable final String query) {
        for (int i = 0; i < mPages.size(); ++i) {
            final WeakReference<SearchableFrameLayout> reference = mPages.get(i);

            if (reference != null) {
                final SearchableFrameLayout view = reference.get();

                if (view != null && view.isAlive()) {
                    view.search(query);
                }
            }
        }
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/adapters/HomePagerAdapter.kt
package com.garpr.android.adapters

import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import com.garpr.android.misc.Heartbeat
import com.garpr.android.misc.HomeTab
import com.garpr.android.misc.ListLayout
import com.garpr.android.misc.RankingCriteriaHandle
import com.garpr.android.misc.Refreshable
import com.garpr.android.misc.Searchable
import com.garpr.android.models.RankingCriteria
import com.garpr.android.models.RankingsBundle
import com.garpr.android.views.FavoritePlayersLayout
import com.garpr.android.views.RankingsLayout
import com.garpr.android.views.TournamentsLayout
import java.lang.ref.WeakReference

class HomePagerAdapter : PagerAdapter(), RankingCriteriaHandle, Refreshable, Searchable {

    private val pages = mutableMapOf<HomeTab, WeakReference<View?>?>()


    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View)
        pages.remove(homeTab(position))
    }

    override fun getCount() = HomeTab.values().size

    private fun homeTab(position: Int): HomeTab {
        return HomeTab.values()[position]
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val homeTab = homeTab(position)
        val view: View = when (homeTab) {
            HomeTab.RANKINGS -> RankingsLayout.inflate(container)
            HomeTab.TOURNAMENTS -> TournamentsLayout.inflate(container)
            HomeTab.FAVORITE_PLAYERS -> FavoritePlayersLayout.inflate(container)
        }

        container.addView(view)
        pages[homeTab] = WeakReference(view)

        return view
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

    fun onNavigationItemReselected(homeTab: HomeTab) {
        val view = pages[homeTab]?.get()

        if ((view as? Heartbeat)?.isAlive == true) {
            (view as? ListLayout)?.smoothScrollToTop()
        }
    }

    val rankingsBundle: RankingsBundle?
        get() {
            val view = pages[HomeTab.RANKINGS]?.get()
            return if ((view as? RankingsLayout)?.isAlive == true) view.rankingsBundle else null
        }

    override val rankingCriteria: RankingCriteria?
        get() = rankingsBundle?.rankingCriteria

    override fun refresh() {
        pages.forEach {
            if ((it.value as? Heartbeat)?.isAlive == true) {
                (it.value as? Refreshable)?.refresh()
            }
        }
    }

    override fun search(query: String?) {
        pages.forEach {
            if ((it.value as? Heartbeat)?.isAlive == true) {
                (it.value as? Searchable)?.search(query)
            }
        }
    }

}
